src/helpers/MeasureBundleHelpers.ts

```diff
@@ -107,6 +107,61 @@ export function extractCompositeMeasure(measureBundle: fhir4.Bundle): fhir4.Meas
   return allCompositeMeasures[0];
 }
 
+/**
+ * Extracts CQFM Group Id from a given composite measure Related Artifact.
+ * https://build.fhir.org/ig/HL7/cqf-measures/StructureDefinition-cqfm-groupId.html
+ * @param relatedArtifact related artifact defined on the composite measure
+ * @returns group id, if defined
+ */
+export function getGroupIdForComponent(relatedArtifact: fhir4.RelatedArtifact): string | null {
+  const groupIdExtension = relatedArtifact.extension?.find(
+    ({ url }) => url === 'http://hl7.org/fhir/us/cqfmeasures/StructureDefinition/cqfm-groupId'
+  );
+  return groupIdExtension?.valueString ?? null;
+}
+
+/**
+ * Filters component results using mapping of group ids defined by the CQFM Group Id extension on the
+ * composite measure. Throws error if a component contains multiple groups but no group is specified to
+ * be used for measure score calculation.
+ * @param componentGroupIds mapping of group ids defined by the CQFM Group Id extension on the composite
+ * @param componentResults array of component results from detailed results
+ * @returns component results, filtered by desired group Ids
+ */
+export function filterComponentResults(
+  componentGroupIds: Record<string, string>,
+  componentResults?: ComponentResults[]
+): ComponentResults[] {
+  const filteredComponentResults: ComponentResults[] = [];
+  componentResults?.forEach(cr => {
```

Comment:
I think that this might be more straightforward using `Array.filter()` rather than pushing things onto `filteredComponentResults using `forEach`...
Probably would not be too big a change. Something like:
```suggestion
  const filteredComponentResults = componentResults?.filter(cr => {
```
... and then instead of pushing cr onto filteredComponentResults, you would return true in those cases.